### PR TITLE
[DI] Throw useful exception on bad XML argument tags

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22525
| License       | MIT
| Doc PR        | -

I still think that the feature request in #22525 would make things better.
But at least, let's make thing fail loudly, instead of silently today, with the associated usual wtfs :)

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Needs #22529 to be green.

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 1]
```diff
@@ -474,6 +476,9 @@ private function getArgumentsAsPhp(\DOMElement $node, $name, $lowercase = true,
 
             switch ($arg->getAttribute('type')) {
                 case 'service':
+                    if (!$arg->getAttribute('id')) {
```
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: `hasAttribute`?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: An empty attribute is also invalid, so better as is IMHO

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 2]
```diff
@@ -484,17 +489,23 @@ private function getArgumentsAsPhp(\DOMElement $node, $name, $lowercase = true,
                     $arguments[$key] = new Expression($arg->nodeValue);
                     break;
                 case 'closure-proxy':
+                    if (!$arg->getAttribute('id')) {
+                        throw new InvalidArgumentException(sprintf('Tag "<%s>" with type="service" is missing an "id" attribute in "%s".', $name, $file));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: wrong type

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 3]
```diff
@@ -484,17 +489,23 @@ private function getArgumentsAsPhp(\DOMElement $node, $name, $lowercase = true,
...
+                    }
+                    if (!$arg->getAttribute('method')) {
+                        throw new InvalidArgumentException(sprintf('Tag "<%s>" with type="service" is missing a "method" attribute in "%s".', $name, $file));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: wrong type

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: fixed thanks

**Comment by nicolas-grekas** [comment]
rebased

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php (hunk 1 above)
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: Ok :+1:

**Review by GuilhemN — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 4]
```diff
@@ -474,6 +476,9 @@ private function getArgumentsAsPhp(\DOMElement $node, $name, $lowercase = true,
 
             switch ($arg->getAttribute('type')) {
                 case 'service':
+                    if (!$arg->getAttribute('id')) {
+                        throw new InvalidArgumentException(sprintf('Tag "<%s>" with type="service" is missing an "id" attribute in "%s".', $name, $file));
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: is missing or is empty?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: fixed

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**